isobus/src/isobus_device_descriptor_object_pool_helpers.cpp

```diff
@@ -0,0 +1,419 @@
+//================================================================================================
+/// @file isobus_device_descriptor_object_pool_helpers.cpp
+///
+/// @brief Implements helpers for the DeviceDescriptorObjectPool class.
+/// @author Adrian Del Grosso
+///
+/// @copyright 2024 The Open-Agriculture Developers
+//================================================================================================
+#include "isobus/isobus/isobus_device_descriptor_object_pool_helpers.hpp"
+
+#include "isobus/isobus/can_stack_logger.hpp"
+#include "isobus/isobus/isobus_standard_data_description_indices.hpp"
+
+namespace isobus
+{
+	DeviceDescriptorObjectPoolHelper::Implement DeviceDescriptorObjectPoolHelper::get_implement_geometry(DeviceDescriptorObjectPool &ddop)
+	{
+		Implement retVal;
+
+		if (0 == ddop.size())
+		{
+			CANStackLogger::error("[DDOP Helper]: No objects in the pool.");
+			return retVal; // Return empty object
+		}
+
+		// First, find the device object
+		for (std::uint32_t i = 0; i < ddop.size(); i++)
+		{
+			auto deviceObject = ddop.get_object_by_index(i);
+
+			if ((nullptr != deviceObject) &&
+			    (task_controller_object::ObjectTypes::Device == deviceObject->get_object_type()))
+			{
+				// Next, iterate through all elements whose parent is the device object
+				for (std::uint32_t j = 0; j < ddop.size(); j++)
+				{
+					auto currentElementObject = ddop.get_object_by_index(j);
+
+					if ((nullptr != currentElementObject) &&
+					    (task_controller_object::ObjectTypes::DeviceElement == currentElementObject->get_object_type()) &&
+					    (std::static_pointer_cast<task_controller_object::DeviceElementObject>(currentElementObject)->get_parent_object() == deviceObject->get_object_id()))
+					{
+						// All the things we care about are likely in here, under the device element.
+						for (std::uint32_t k = 0; k < ddop.size(); k++)
+						{
+							auto potentialFunction = ddop.get_object_by_index(k);
+
+							if ((nullptr != potentialFunction) &&
+							    (task_controller_object::ObjectTypes::DeviceElement == potentialFunction->get_object_type()) &&
+							    (std::static_pointer_cast<task_controller_object::DeviceElementObject>(potentialFunction)->get_parent_object() == currentElementObject->get_object_id()) &&
+							    (task_controller_object::DeviceElementObject::Type::Function == std::static_pointer_cast<task_controller_object::DeviceElementObject>(potentialFunction)->get_type()))
```

Comment:
See ISO11783-10 part F.3.1, a boom can both be either a function OR a device. So I think requiring an object to be a function is not always correct?

Reply:
Without a device, object the DDOP wouldn't be very functional, so it maybe makes sense to check for a function first, then if none is found, consider the device to be the root of the boom?

Reply:
Added this functionality and added a unit test specific to this case.